File: shared/py/baseApp.py
"""
Base FastAPI Application
Provides common setup and utilities for FastAPI services
"""
from fastapi import FastAPI, Request, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.middleware.trustedhost import TrustedHostMiddleware
from fastapi.responses import JSONResponse
import logging
import os
import sys
import uvicorn
import time
import psutil
from typing import Optional, Dict, Any, Callable
from contextlib import asynccontextmanager

# Add shared/py to path for imports
_shared_py_path = os.path.dirname(os.path.abspath(__file__))
if _shared_py_path not in sys.path:
    sys.path.insert(0, _shared_py_path)

from redisManager import RedisManager
from dbManager import DBManager
from queueManager import QueueManager
from emailService import EmailService
from webSocketServer import WebSocketServer


class BaseApp:
    """Base FastAPI application with common utilities"""

    def __init__(
        self,
        name: str,
        version: str = "1.0.0",
        custom_lifespan: Optional[Callable] = None,
        custom_cors: Optional[Dict[str, Any]] = None,
    ):
        self.name = name
        self.version = version
        self.custom_lifespan = custom_lifespan
        self.start_time = time.time()
        
        # Initialize managers
        self.redis_manager: Optional[RedisManager] = None
        self.db_manager: Optional[DBManager] = None
        self.queue_manager: Optional[QueueManager] = None
        self.email_service: Optional[EmailService] = None
        self.websocket_server: Optional[WebSocketServer] = None

        # Setup logging
        self._setup_logging()

        # Create FastAPI app
        lifespan_context = custom_lifespan if custom_lifespan else self._lifespan
        self.app = FastAPI(
            title=name,
            version=version,
            docs_url="/docs",
            redoc_url="/redoc",
            lifespan=lifespan_context,
        )

        # Setup middleware
        self._setup_middleware(custom_cors)

        # Setup routes
        self._setup_routes()

        # Error handling will be initialized separately via initialize_error_handling()

    def _setup_logging(self):
        """Setup application logging"""
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        self.logger = logging.getLogger(self.name)

    def _setup_middleware(self, custom_cors: Optional[Dict[str, Any]] = None):
        """Setup FastAPI middleware"""
        # CORS middleware
        cors_config = custom_cors or {
            "allow_origins": ["*"],
            "allow_credentials": True,
            "allow_methods": ["*"],
            "allow_headers": ["*"],
        }
        
        self.app.add_middleware(
            CORSMiddleware,
            **cors_config
        )

        # Trusted host middleware (for production)
        if os.getenv("ENV") == "production":
            self.app.add_middleware(
                TrustedHostMiddleware,
                allowed_hosts=["*"]
            )

    def _setup_routes(self):
        """Setup common routes"""
        @self.app.get("/health")
        async def health_check():
            """Health check endpoint with Prometheus-style metrics"""
            # Get memory usage
            memory_info = psutil.virtual_memory()
            process = psutil.Process()
            process_memory = process.memory_info()
            
            return {
                "status": "OK",
                "service": self.name,
                "version": self.version,
                "timestamp": time.time(),
                "uptime": time.time() - self.start_time,
                "memory": {
                    "rss": process_memory.rss,
                    "vms": process_memory.vms,
                    "percent": process.memory_percent(),
                    "system_total": memory_info.total,
                    "system_available": memory_info.available,
                    "system_percent": memory_info.percent
                },
                "cpu": {
                    "percent": psutil.cpu_percent(interval=0.1)
                }
            }

        @self.app.get("/")
        async def root():
            """Root endpoint"""
            return {
                "message": f"Welcome to {self.name}",
                "version": self.version,
                "docs": "/docs"
            }

    @asynccontextmanager
    async def _lifespan(self, app: FastAPI):
        """Application lifespan context manager"""
        # Startup
        await self._startup()
        yield
        # Shutdown
        await self._shutdown()

    async def _startup(self):
        """Application startup logic"""
        self.logger.info(f"Starting {self.name} v{self.version}")

        # Initialize Redis if configured
        if os.getenv("REDIS_HOST"):
            self.redis_manager = RedisManager()
            await self.redis_manager.connect()

        # Initialize DB if configured
        db_url = os.getenv("DATABASE_URL") or os.getenv("DB_URL")
        if db_url:
            self.db_manager = DBManager.get_instance(database_url=db_url)
            await self.db_manager.connect()

        # Initialize Queue Manager if Redis is available
        if self.redis_manager:
            self.queue_manager = QueueManager(self.redis_manager)

        # Initialize Email Service if configured
        if os.getenv("EMAIL_HOST"):
            self.email_service = EmailService()

        # Initialize WebSocket Server if needed
        if os.getenv("ENABLE_WEBSOCKETS", "false").lower() == "true":
            self.websocket_server = WebSocketServer(self.redis_manager)

    async def _shutdown(self):
        """Application shutdown logic"""
        self.logger.info(f"Shutting down {self.name}")

        # Close connections
        if self.redis_manager:
            await self.redis_manager.disconnect()

        if self.db_manager:
            await self.db_manager.disconnect()

        if self.websocket_server:
            await self.websocket_server.disconnect()

    def initialize_error_handling(self):
        """Initialize error handling middleware"""
        # Import error handler from middleware
        try:
            middleware_path = os.path.join(
                os.path.dirname(__file__), "../../middleware/py"
            )
            if middleware_path not in sys.path:
                sys.path.insert(0, middleware_path)
            
            from error_handler import error_handler
            
            async def exception_handler(request: Request, exc: Exception):
                return await error_handler(request, exc)
            
            self.app.add_exception_handler(Exception, exception_handler)
            self.logger.info("Error handling initialized")
        except ImportError as e:
            self.logger.warning(f"Error handler not found, using default: {e}")
            # Fallback to default error handler
            self.app.add_exception_handler(
                Exception,
                lambda request, exc: self.handle_error(request, exc)
            )

    def add_routes(self, path: str, router):
        """Add routes to the application"""
        self.app.include_router(router, prefix=path)

    async def start(self, db: Optional[DBManager] = None, port: int = 8000):
        """Start the server"""
        if db:
            await db.connect()
            self.db_manager = db
        elif not self.db_manager:
            db_url = os.getenv("DATABASE_URL") or os.getenv("DB_URL")
            if db_url:
                self.db_manager = DBManager.get_instance(database_url=db_url)
                await self.db_manager.connect()

        # Initialize Redis if configured
        if os.getenv("REDIS_HOST") and not self.redis_manager:
            self.redis_manager = RedisManager()
            await self.redis_manager.connect()

        # Initialize Queue Manager if Redis is available
        if self.redis_manager and not self.queue_manager:
            self.queue_manager = QueueManager(self.redis_manager)

        self.logger.info(f"🚀 {self.name} starting on port {port}")
        config = uvicorn.Config(
            app=self.app,
            host="0.0.0.0",
            port=port,
            log_level="info",
        )
        server = uvicorn.Server(config)
        await server.serve()

    async def shutdown(self, db: Optional[DBManager] = None):
        """Graceful shutdown"""
        self.logger.info(f"🛑 Shutting down {self.name}")

        if self.queue_manager:
            await self.queue_manager.shutdown()

        if self.redis_manager:
            await self.redis_manager.disconnect()

        if self.db_manager:
            await self.db_manager.disconnect()
        elif db:
            await db.disconnect()

        if self.websocket_server:
            await self.websocket_server.disconnect()

        self.logger.info(f"✅ {self.name} shutdown complete")

    def get_app(self) -> FastAPI:
        """Get the FastAPI application instance"""
        return self.app

    async def handle_error(self, request: Request, exc: Exception) -> JSONResponse:
        """Default error handler fallback"""
        self.logger.error(f"Error processing request: {exc}")

        if isinstance(exc, HTTPException):
            return JSONResponse(
                status_code=exc.status_code,
                content={"detail": exc.detail}
            )

        return JSONResponse(
            status_code=500,
            content={"detail": "Internal server error"}
        )

File: shared/js/baseApp.ts
import express, { Application, Request, Response, NextFunction } from "express";
import morgan from "morgan";
import helmet from "helmet";
import cookieParser from "cookie-parser";
import compression from "compression";
import cors from "cors";
import createHttpError from "http-errors";
import rateLimit from "express-rate-limit";
import http from "http";
import cluster from "cluster";
import os from "os";
import { QueueManager } from "./queueManager";
import { RedisManager } from "./redisManager";

import { errorHandler } from "../../middleware/js/errorHandler";

// Helper function to sanitize objects (prevent NoSQL injection)
function sanitizeObject(obj: any): void {
  if (!obj || typeof obj !== "object") return;
  
  for (const key in obj) {
    if (key.includes("$") || key.includes(".")) {
      delete obj[key];
    } else if (typeof obj[key] === "object" && obj[key] !== null) {
      sanitizeObject(obj[key]);
    }
  }
}
import { BaseConfig } from "../../types/config";
import webSocketService from "./webSocketServer";
import { DatabaseManager } from "./dbManager";

export interface AppOptions {
  serviceName: string;
  config: BaseConfig;
  enableSockets?: boolean;
  enableSessions?: boolean;
  enableFileUpload?: boolean;
  enableQueues?: boolean;
  disableCors?: boolean; // new option, default false
  customRateLimit?: {
    windowMs: number;
    max: number;
  };
  customCors?: cors.CorsOptions;
}

export class BaseApp {
  public app: Application;
  public server?: http.Server;
  public io?: any; // Server from socket.io (optional)
  public queueManager?: QueueManager;
  private config: BaseConfig;
  private serviceName: string;

  constructor(options: AppOptions) {
    this.app = express();
    this.config = options.config;
    this.serviceName = options.serviceName;

    this.initializeMiddleware(options);

    if (options.enableSockets) {
      this.initializeSocketIO();
    }

    if (options.enableQueues) {
      this.initializeQueues();
    }
  }

  private initializeMiddleware(options: AppOptions): void {
    // Trust proxy for deployment
    this.app.set("trust proxy", 1);

    // Development logging
    if (this.config.NODE_ENV !== "production") {
      this.app.use(morgan("dev"));
    }

    // Security middleware
    this.app.disable("x-powered-by");
    this.app.use(
      helmet({
        contentSecurityPolicy: false,
      })
    );

    // Rate limiting
    const rateLimitConfig = options.customRateLimit || {
      windowMs: 15 * 60 * 1000, // 15 minutes
      max: 100,
    };

    const limiter = rateLimit({
      ...rateLimitConfig,
      message: "Too many requests from this IP, please try again later.",
      standardHeaders: true,
      legacyHeaders: false,
    });
    this.app.use("/api/", limiter);

    // CORS configuration
    const corsOptions: cors.CorsOptions = options.customCors || {
      origin: this.config.CLIENT_URL,
      credentials: true,
      optionsSuccessStatus: 200,
      methods: ["GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS"],
      allowedHeaders: ["Content-Type", "Authorization", "x-api-key"],
    };
    // Apply CORS only if not explicitly disabled (default: enabled)
    if (!options.disableCors) {
      this.app.use(cors(corsOptions));
    }

    // Body parsing middleware
    this.app.use(cookieParser() as express.RequestHandler);
    // Parse JSON bodies first
    this.app.use(express.json({ limit: "10mb" }) as express.RequestHandler);
    // Parse URL-encoded bodies
    this.app.use(
      express.urlencoded({ extended: true, limit: "10mb" }) as express.RequestHandler
    );

    // Session middleware (optional - dynamically imported)
    if (options.enableSessions) {
      this.initializeSessions();
    }

    // File upload (optional - dynamically imported)
    if (options.enableFileUpload) {
      this.initializeFileUpload();
    }

    // Health check endpoint
    this.app.get("/health", (req: Request, res: Response) => {
      res.status(200).json({
        status: "OK",
        service: this.serviceName,
        timestamp: new Date().toISOString(),
        uptime: process.uptime(),
        memory: process.memoryUsage(),
      });
    });

    // Security and optimization - basic input sanitization
    this.app.use("/", (req: any, res: any, next: any) => {
      // Basic sanitization - remove $ and . from object keys to prevent NoSQL injection
      if (req.body && typeof req.body === "object") {
        sanitizeObject(req.body);
      }
      if (req.params && typeof req.params === "object") {
        sanitizeObject(req.params);
      }
      next();
    });
    this.app.use(compression());

    // Attach socket.io to request object for route handlers (if enabled)
    if (this.io) {
      this.app.use("/", (req: any, res: any, next: any) => {
        req.io = this.io;
        next();
      });
    }
  }

  private initializeSessions(): void {
    try {
      // Dynamically import express-session only if needed
      const session = require("express-session");
      
      // Try to use Redis store if available, otherwise use memory store
      let store: any;
      try {
        const RedisStore = require("connect-redis").default;
        const { createClient } = require("redis");
        const redisClient = createClient({
          url: process.env.REDIS_URL || "redis://localhost:6379",
        });
        redisClient.connect().catch(console.error);
        store = new RedisStore({ client: redisClient });
        console.log("✅ Using Redis for session store");
      } catch (redisError) {
        // Fallback to memory store if Redis is not available
        try {
          const MemoryStore = require("memorystore")(session);
          store = new MemoryStore({
            checkPeriod: 86400000, // prune expired entries every 24h
          });
          console.log("✅ Using memory store for sessions");
        } catch (memoryError) {
          // If memorystore is not available, use default memory store
          store = undefined;
          console.log("⚠️  Using default memory store (sessions lost on restart)");
        }
      }

      const sessionConfig: any = {
        secret: this.config.SESSION_SECRET || "default-secret",
        resave: false,
        saveUninitialized: false,
        store,
        cookie: {
          secure: this.config.NODE_ENV === "production",
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 7, // 7 days
          sameSite: this.config.NODE_ENV === "production" ? "none" : "lax",
        },
      };

      this.app.use(session(sessionConfig));
      console.log("✅ Sessions initialized");
    } catch (error) {
      console.warn(
        "⚠️  express-session not installed. Sessions disabled."
      );
      console.warn("   Install express-session if you need sessions.");
    }
  }

  private initializeFileUpload(): void {
    try {
      // Dynamically import express-fileupload only if needed
      const fileUpload = require("express-fileupload");

      this.app.use(
        "/",
        fileUpload({
          useTempFiles: true,
          tempFileDir: "/tmp/",
          limits: { fileSize: 50 * 1024 * 1024 },
        }) as unknown as express.RequestHandler
      );
      console.log("✅ File upload initialized");
    } catch (error) {
      console.warn(
        "⚠️  express-fileupload not installed. File upload disabled."
      );
      console.warn("   Install express-fileupload if you need file upload support.");
    }
  }

  private initializeSocketIO(): void {
    try {
      // Dynamically import socket.io only if needed
      const { Server } = require("socket.io");
      this.server = http.createServer(this.app);
      this.io = new Server(this.server, {
        pingTimeout: 60000,
        cors: {
          origin: this.config.CLIENT_URL,
          credentials: true,
        },
      });

      // Attach socket.io to request object for route handlers
      this.app.use((req: any, res: any, next: any) => {
        req.io = this.io;
        next();
      });

      // Initialize global WebSocket service with io instance
      webSocketService.initializeWithIO(this.io);

      // Optionally: Make globally available
      (global as any).io = this.io;
      (global as any).webSocketService = webSocketService;
    } catch (error) {
      console.warn("⚠️  socket.io not installed. WebSocket features disabled.");
      console.warn("   Install socket.io if you need WebSocket support.");
    }
  }

  private async initializeQueues(): Promise<void> {
    this.queueManager = QueueManager.getInstance();
    await this.queueManager.initialize();
    (global as any).queueManager = this.queueManager;
  }

  // Method to add routes
  public addRoutes(path: string, router: express.Router): void {
    this.app.use(path, router);
  }
  public initializeErrorHandling(): void {
    // 404 handler
    this.app.use("/", ((req: Request, res: Response, next: NextFunction) => {
      next(createHttpError.NotFound(`Route ${req.originalUrl} not found`));
    }) as express.RequestHandler);

    // Global error handler
    this.app.use("/", errorHandler as unknown as express.ErrorRequestHandler);
  }

  // Method to start the server
  public listen(port: number): Promise<void> {
    return new Promise((resolve) => {
      const serverInstance = this.server || this.app;

      serverInstance.listen(port, () => {
        console.log(`🚀 ${this.serviceName} listening on port ${port}`);
        resolve();
      });
    });
  }

  // Method to add socket event handlers
  public addSocketHandlers(handler: (io: any) => void): void {
    if (this.io) {
      handler(this.io);
    } else {
      throw new Error(
        "Socket.IO not initialized. Set enableSockets: true in options."
      );
    }
  }

  // Graceful shutdown
  public async close(): Promise<void> {
    if (this.server) {
      return new Promise((resolve) => {
        this.server!.close(() => {
          console.log(`🛑 ${this.serviceName} server closed`);
          resolve();
        });
      });
    }
  }

  public async start(db: DatabaseManager | null, port: number) {
    const useCluster = process.env.USE_CLUSTER === "true";

    if (useCluster && cluster.isPrimary) {
      // Primary process - fork workers
      await this.startAsPrimary(db, port);
    } else {
      // Worker process or non-clustered mode
      await this.startAsWorker(db, port);
    }
  }

  private async startAsPrimary(db: DatabaseManager | null, port: number) {
    const numCPUs = os.availableParallelism();
    const maxWorkers = process.env.MAX_WORKERS
      ? parseInt(process.env.MAX_WORKERS, 10)
      : numCPUs;

    console.log(
      `🚀 ${this.serviceName} Primary process ${process.pid} starting`
    );
    console.log(
      `📊 Detected ${numCPUs} CPU cores, spawning ${maxWorkers} workers`
    );

    // Fork workers
    for (let i = 0; i < maxWorkers; i++) {
      const worker = cluster.fork();
      console.log(
        `👷 Worker ${worker.id} (PID: ${worker.process.pid}) spawned`
      );
    }

    // Handle worker lifecycle events
    cluster.on("online", (worker) => {
      console.log(
        `✅ Worker ${worker.id} (PID: ${worker.process.pid}) is online`
      );
    });

    cluster.on("listening", (worker, address) => {
      console.log(
        `🎧 Worker ${worker.id} (PID: ${worker.process.pid}) listening on ${address.address}:${address.port}`
      );
    });

    cluster.on("disconnect", (worker) => {
      console.log(
        `⚠️  Worker ${worker.id} (PID: ${worker.process.pid}) disconnected`
      );
    });

    cluster.on("exit", (worker, code, signal) => {
      if (worker.exitedAfterDisconnect === true) {
        console.log(
          `👋 Worker ${worker.id} (PID: ${worker.process.pid}) exited voluntarily`
        );
      } else {
        console.error(
          `💥 Worker ${worker.id} (PID: ${worker.process.pid}) died unexpectedly (${signal || code}). Restarting...`
        );

        // Respawn the worker
        const newWorker = cluster.fork();
        console.log(
          `🔄 New worker ${newWorker.id} (PID: ${newWorker.process.pid}) spawned as replacement`
        );
      }
    });

    // Graceful shutdown handler for primary
    const shutdownPrimary = async () => {
      console.log(`\n🛑 ${this.serviceName} Primary received shutdown signal`);
      console.log("📢 Gracefully shutting down all workers...");

      const workers = Object.values(cluster.workers || {});
      const shutdownPromises = workers.map((worker) => {
        return new Promise<void>((resolve) => {
          if (!worker) {
            resolve();
            return;
          }

          const timeout = setTimeout(() => {
            console.log(`⏰ Worker ${worker.id} timeout, forcing kill`);
            worker.kill("SIGKILL");
            resolve();
          }, 10000); // 10 second timeout

          worker.on("exit", () => {
            clearTimeout(timeout);
            resolve();
          });

          worker.disconnect();
          worker.kill("SIGTERM");
        });
      });

      await Promise.all(shutdownPromises);
      console.log("✅ All workers shut down successfully");
      process.exit(0);
    };

    process.on("SIGTERM", shutdownPrimary);
    process.on("SIGINT", shutdownPrimary);
  }

  private async startAsWorker(db: DatabaseManager | null, port: number) {
    try {
      // Connect to database if provided
      if (db) {
        await db.connect();
      }

      // Note: Redis connection is handled by QueueManager.initialize() if queues are enabled
      // Sessions also handle Redis connection dynamically if needed
      // No need to connect Redis here separately

      await this.listen(port);

      if (cluster.isWorker) {
        console.log(
          `🚀 ${this.serviceName} Worker ${cluster.worker?.id} (PID: ${process.pid}) started successfully`
        );
      } else {
        console.log(
          `🚀 ${this.serviceName} started successfully (PID: ${process.pid})`
        );
      }
    } catch (error) {
      console.error(`💥 Failed to start ${this.serviceName}:`, error);
      process.exit(1);
    }
  }

  public async shutdown(db: DatabaseManager | null) {
    const isWorker = cluster.isWorker;
    const processInfo = isWorker
      ? `Worker ${cluster.worker?.id} (PID: ${process.pid})`
      : `Process (PID: ${process.pid})`;

    console.log(`🛑 Shutting down ${this.serviceName} ${processInfo}...`);

    if (this.queueManager) {
      await this.queueManager.shutdown();
    }

    // Disconnect Redis
    const redisManager = RedisManager.getInstance();
    await redisManager.disconnect();

    await this.close();
    
    // Disconnect database if provided
    if (db) {
      await db.disconnect();
    }

    console.log(`✅ ${this.serviceName} ${processInfo} shutdown complete`);
    process.exit(0);
  }
}
